### PR TITLE
Atexit model implementation

### PR DESCRIPTION
**Pull request by mikhailramalho**
This PR implements the atexit model and changes ESBMC to always call it when defined by the standard.

The biggest change is that finish_formula now call assertion(), instead of creating the assertion itself. This is necessary so we can encode the memory cleanup check using the appropriate guards.

### CODE REVIEW

**Comment by mikhailramalho** [comment]
Some design decisions about the implementation: 
1. I decided to use a fixed array instead of a dynamic one (like the pthread destructor key) to make the model faster. I don't know how much slower a dynamic allocation array would perform.
2. We are fixed to 32 functions, since this is the minimum requirement.

**Review by lucasccordeiro — APPROVED** [review]
LGTM once we get the SV-COMP results from this PR.

**Comment by mikhailramalho** [comment]
I found some performance regressions because the k-induction code havocs the new variables `unsigned short atexit_index = 0;` and `void (*atexit_func[32])();`.

A dynamically allocated array should fix this...

File: src/clang-c-frontend/clang_c_language.cpp [hunk 1]
```diff
@@ -319,12 +319,16 @@ int __ESBMC_rounding_mode = 0;
 
 void *__ESBMC_memset(void *, int, unsigned int);
 
+void __ESBMC_finish_formula();
```
Comment (mikhailramalho) [inline src/clang-c-frontend/clang_c_language.cpp]: this is probably not needed.

File: src/clang-c-frontend/clang_c_language.cpp [hunk 2]
```diff
@@ -319,12 +319,16 @@ int __ESBMC_rounding_mode = 0;
...
+
 // Forward decs for pthread main thread begin/end hooks. Because they're
 // pulled in from the C library, they need to be declared prior to pulling
 // them in, for type checking.
 void pthread_start_main_hook(void);
 void pthread_end_main_hook(void);
 
+void __atexit_handler(void);
```
Comment (mikhailramalho) [inline src/clang-c-frontend/clang_c_language.cpp:330]: can we get rid of it? Since it is added in clang_c_main?

File: src/util/simplify_expr2.cpp [hunk 3]
```diff
@@ -1912,7 +1912,27 @@ static expr2tc simplify_relations(
...
+          return true;
+      }
+      return false;
+    };
+
+    std::function<int(expr2tc &)> get_value = [](expr2tc &) -> int {
+      return 0xbadbeef;
+    };
+
+    simpl_res = TFunctor<int>::simplify(
```
Comment (rafaelsamenezes) [inline src/util/simplify_expr2.cpp:1935]: Shoudn't this be moved to another PR?

File: src/goto-symex/symex_main.cpp [hunk 4]
```diff
@@ -703,17 +707,8 @@ void goto_symext::finish_formula()
...
     cur_state->rename(eq);
-    target->assertion(
-      it.alloc_guard.as_expr(),
-      eq,
-      "dereference failure: forgotten memory: " + get_pretty_name(it.name),
-      cur_state->gen_stack_trace(),
-      cur_state->source,
-      first_loop);
-
-    total_claims++;
```
Comment (rafaelsamenezes) [inline src/goto-symex/symex_main.cpp:716]: `goto_symext::assertion` does not add to total_claims.

File: src/c2goto/library/stdlib.c [hunk 5]
```diff
@@ -8,15 +8,65 @@
...
+__ESBMC_HIDE:;
+  __atexit_handler();
+  __ESBMC_finish_formula();
   __ESBMC_assume(0);
 }
 
 void abort(void)
 {
+__ESBMC_HIDE:;
+  __atexit_handler();
```
Comment (rafaelsamenezes) [inline src/c2goto/library/stdlib.c]: Not sure about this.  From [Linux Die](https://linux.die.net/man/3/atexit)
```
The atexit() function registers the given function to be called at normal process termination, either via [exit](https://linux.die.net/man/3/exit)(3) or via return from the program's main().
```
I.e., abort should not call the atexit functions.

Did you do this for sv-comp?

File: src/clang-c-frontend/clang_c_language.cpp (hunk 1 above)
Comment (rafaelsamenezes) [reply src/clang-c-frontend/clang_c_language.cpp]: I am not against having an intrinsic that makes the symex stop the current path. There is a use case that is different from `assume(0)`

```c
// some code
assume(0);
while(1)
 // some code
```

Is different from having an intrinsic that just stop the analysis. But we can move this to another PR

**Review by rafaelsamenezes — CHANGES_REQUESTED** [review]

File: src/util/simplify_expr2.cpp (hunk 3 above)
Comment (mikhailramalho) [reply src/util/simplify_expr2.cpp:1935]: This is needed to prevent a performance regression! 
```
void __atexit_handler()
{
__ESBMC_HIDE:;
  // This is here to prevent k-induction from unwind the next loop unnecessarily
  if(__ESBMC_atexits == NULL)
    return;

  while(__ESBMC_atexits)
  {
    __ESBMC_atexits->atexit_func();
    __ESBMC_atexit_key *__ESBMC_tmp = __ESBMC_atexits->next;
    free(__ESBMC_atexits);
    __ESBMC_atexits = __ESBMC_tmp;
  }
}
```
Without it, we go ahead and try to unwind the loop

File: src/goto-symex/symex_main.cpp (hunk 4 above)
Comment (mikhailramalho) [reply src/goto-symex/symex_main.cpp:716]: `goto_symext::assertion` seems broken then.

File: src/c2goto/library/stdlib.c (hunk 5 above)
Comment (mikhailramalho) [reply src/c2goto/library/stdlib.c]: actually, I'm not sure why I added that... I think we assume abort is a valid program termination in sv-comp? I guess we can remove that and re run the sv-comp and see if we get any regression.

**Comment by mikhailramalho** [comment]
> I am not against having an intrinsic that makes the symex stop the current path. There is a use case that is different from assume(0)
> 
> // some code
> assume(0);
> while(1)
>  // some code
> Is different from having an intrinsic that just stop the analysis. But we can move this to another PR

Actually, this intrinsic only call symex::finish_formula which (despite its name) doesn't finish the formula per se, it only adds memory-leak checks, if it is enabled of course.

File: src/util/simplify_expr2.cpp (hunk 3 above)
Comment (rafaelsamenezes) [reply src/util/simplify_expr2.cpp:1935]: I see, could you update your comment to reflect this then? Also, could you add testcases with `--no-simplify`?

**Comment by rafaelsamenezes** [comment]
> Actually, this intrinsic only call symex::finish_formula which (despite its name) doesn't finish the formula per se, it only adds memory-leak checks, if it is enabled of course.

I see, we should do a refactoring then. It goes beyond this PR though.

**Comment by mikhailramalho** [comment]
> > Actually, this intrinsic only call symex::finish_formula which (despite its name) doesn't finish the formula per se, it only adds memory-leak checks, if it is enabled of course.
> 
> I see, we should do a refactoring then. It goes beyond this PR though.

yeah, although the documentation clearly states what this function is supposed to do, I think changing its name is desirable:
```
  /**
   *  Perform accounting checks / assertions at end of a program run.
   *  This should contain anything that must happen at the end of a program run,
   *  for example assertions about dynamic memory being freed.
   */
  void finish_formula();
```

File: src/util/simplify_expr2.cpp (hunk 3 above)
Comment (mikhailramalho) [reply src/util/simplify_expr2.cpp:1935]: sure!

**Comment by mikhailramalho** [comment]
@rafaelsamenezes could you take another look? I implemented all you comments.

**Review by rafaelsamenezes — APPROVED** [review]
